The Simple Recommender offers generalized recommnendations to every user based on movie popularity and (sometimes) genre. The basic idea behind this recommender is that movies that are more popular and more critically acclaimed will have a higher probability of being liked by the average audience. This model does not give personalized recommendations based on the user.

In [1]:
import pandas as pd
import numpy as np
from ast import literal_eval

In [41]:
md = pd.read_csv('../0_data/input/movies_metadata.csv')

In [42]:
md

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45458  False                                                NaN         0   
45459  False                                                NaN         0   
45460  False                                                NaN         0   
45461  False                                                NaN         0   
45462  False                                                NaN         0   

                                   genres  \
0       ['Animation', 'Comedy', 'Family']   
1      ['Adventure', 'Fantasy', 'Family']   
2                   ['Romance', 'Comedy']   
3          ['Comedy', 'Drama', 'Romance']   
4                              ['Comedy']   
...                                   ...   
45458                 ['Drama', 'Family']   
45459                           ['Drama']   
45460     ['Action', 'Drama', 'Thriller']   
45461                                  []   
45462                                  []   

                                   homepage  tmdbId    imdb_id  \
0      http://toystory.disney.com/toy-story     862  tt0114709   
1                                       NaN    8844  tt0113497   
2                                       NaN   15602  tt0113228   
3                                       NaN   31357  tt0114885   
4                                       NaN   11862  tt0113041   
...                                     ...     ...        ...   
45458  http://www.imdb.com/title/tt6209470/  439050  tt6209470   
45459                                   NaN  111109  tt2028550   
45460                                   NaN   67758  tt0303758   
45461                                   NaN  227506  tt0008536   
45462                                   NaN  461257  tt6980792   

      original_language               original_title  \
0                    en                    Toy Story   
1                    en                      Jumanji   
2                    en             Grumpier Old Men   
3                    en            Waiting to Exhale   
4                    en  Father of the Bride Part II   
...                 ...                          ...   
45458                fa                      رگ خواب   
45459                tl          Siglo ng Pagluluwal   
45460                en                     Betrayal   
45461                en          Satana likuyushchiy   
45462                en                     Queerama   

                                                overview  ...  release_date  \
0      Led by Woody, Andy's toys live happily in his ...  ...    1995-10-30   
1      When siblings Judy and Peter discover an encha...  ...    1995-12-15   
2      A family wedding reignites the ancient feud be...  ...    1995-12-22   
3      Cheated on, mistreated and stepped on, the wom...  ...    1995-12-22   
4      Just when George Banks has recovered from his ...  ...    1995-02-10   
...                                                  ...  ...           ...   
45458        Rising and falling between a man and woman.  ...           NaN   
45459  An artist struggles to finish his work while a...  ...    2011-11-17   
45460  When one of her hits goes wrong, a professiona...  ...    2003-08-01   
45461  In a small town live two brothers, one a minis...  ...    1917-10-21   
45462  50 years after decriminalisation of homosexual...  ...    2017-06-09   

           revenue runtime                                   spoken_languages  \
0      373554033.0    81.0           [{'iso_6

In [43]:
md.dtypes

adult                       bool
belongs_to_collection     object
budget                     int64
genres                    object
homepage                  object
tmdbId                     int64
imdb_id                   object
original_language         object
original_title            object
overview                  object
popularity               float64
poster_path               object
production_companies      object
production_countries      object
release_date              object
revenue                  float64
runtime                  float64
spoken_languages          object
status                    object
tagline                   object
title                     object
video                     object
vote_average             float64
vote_count               float64
dtype: object

We use the TMDB Ratings to come up with our Top Movies Chart. I will use IMDB's weighted rating formula to construct my chart. Mathematically, it is represented as follows:

Weighted Rating (WR) = $(\frac{v}{v+m}\cdot R)+(\frac{m}{v+m}\cdot C)$
 
where,

- v is the number of votes for the movie
- m is the minimum votes required to be listed in the chart
- R is the average rating of the movie
- C is the mean vote across the whole report

The next step is to determine an appropriate value for m, the minimum votes required to be listed in the chart. We will use 95th percentile as our cutoff. In other words, for a movie to feature in the charts, it must have more votes than at least 95% of the movies in the list.


In [44]:
vote_counts = md[md['vote_count'].notnull()]['vote_count'].astype('int')
vote_averages = md[md['vote_average'].notnull()]['vote_average'].astype('int')
C = vote_averages.mean()
C

5.244896612406511

In [45]:
m = vote_counts.quantile(0.95)
m

434.0

In [46]:
md['year'] = pd.to_datetime(md['release_date'], errors='coerce').apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

In [47]:
qualified = md[(md['vote_count'] >= m) & (md['vote_count'].notnull()) & (md['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity', 'genres']]
qualified['vote_count'] = qualified['vote_count'].astype('int')
qualified['vote_average'] = qualified['vote_average'].astype('int')
qualified.shape

(2274, 6)

Therefore, to qualify to be considered for the chart, a movie has to have at least 434 votes on TMDB. We also see that the average rating for a movie on TMDB is 5.244 on a scale of 10. 2274 Movies qualify to be on our chart.

In [48]:
def weighted_rating(x):
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) + (m/(m+v) * C)

In [49]:
qualified['wr'] = qualified.apply(weighted_rating, axis=1)

In [50]:
qualified = qualified.sort_values('wr', ascending=False).head(250)

Top Movies

In [51]:
qualified.head(15)

title  year  vote_count  \
15480                                          Inception  2010       14075   
12481                                    The Dark Knight  2008       12269   
22878                                       Interstellar  2014       11187   
2843                                          Fight Club  1999        9678   
4863   The Lord of the Rings: The Fellowship of the Ring  2001        8892   
292                                         Pulp Fiction  1994        8670   
314                             The Shawshank Redemption  1994        8358   
7000       The Lord of the Rings: The Return of the King  2003        8226   
351                                         Forrest Gump  1994        8147   
5814               The Lord of the Rings: The Two Towers  2002        7641   
256                                            Star Wars  1977        6778   
1225                                  Back to the Future  1985        6239   
834                                        The Godfather  1972        6024   
1154                             The Empire Strikes Back  1980        5998   
46                                                 Se7en  1995        5915   

       vote_average  popularity  \
15480             8   29.108149   
12481             8  123.167259   
22878             8   32.213481   
2843              8   63.869599   
4863              8   32.070725   
292               8  140.950236   
314               8   51.645403   
7000              8   29.324358   
351               8   48.307194   
5814              8   29.423537   
256               8   42.149697   
1225              8   25.778509   
834               8   41.109264   
1154              8   19.470959   
46                8   18.457430   

                                                  genres        wr  
15480  ['Action', 'Thriller', 'Science Fiction', 'Mys...  7.917588  
12481           ['Drama', 'Action', 'Crime', 'Thriller']  7.905871  
22878          ['Adventure', 'Drama', 'Science Fiction']  7.897107  
2843                                           ['Drama']  7.881753  
4863                  ['Adventure', 'Fantasy', 'Action']  7.871787  
292                                ['Thriller', 'Crime']  7.868660  
314                                   ['Drama', 'Crime']  7.864000  
7000                  ['Adventure', 'Fantasy', 'Action']  7.861927  
351                       ['Comedy', 'Drama', 'Romance']  7.860656  
5814                  ['Adventure', 'Fantasy', 'Action']  7.851924  
256           ['Adventure', 'Action', 'Science Fiction']  7.834205  
1225   ['Adventure', 'Comedy', 'Science Fiction', 'Fa...  7.820813  
834                                   ['Drama', 'Crime']  7.814847  
1154          ['Adventure', 'Action', 'Science Fiction']  7.814099  
46                      ['Crime', 'Mystery', 'Thriller']  7.811669

We see that three Christopher Nolan Films, Inception, The Dark Knight and Interstellar occur at the very top of our chart. The chart also indicates a strong bias of TMDB Users towards particular genres and directors.

Let us now construct our function that builds charts for particular genres. For this, we will use relax our default conditions to the 85th percentile instead of 95.

In [69]:
# s = md.apply(lambda x: pd.Series(x['genres']),axis=1).stack().reset_index(level=1, drop=True)
# s.name = 'genre'
# gen_md = md.drop('genres', axis=1).join(s)

In [70]:
# gen_md

In [74]:
def build_chart(genre, percentile=0.85):
    df = md[md['genres'].str.find(genre) >= 0]
    vote_counts = df[df['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = df[df['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(percentile)
    
    qualified = df[(df['vote_count'] >= m) & (df['vote_count'].notnull()) & (df['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity']]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    
    qualified['wr'] = qualified.apply(lambda x: (x['vote_count']/(x['vote_count']+m) * x['vote_average']) + (m/(m+x['vote_count']) * C), axis=1)
    qualified = qualified.sort_values('wr', ascending=False).head(250)
    
    return qualified

Let us see our method in action by displaying the Top 15 Romance Movies (Romance almost didn't feature at all in our Generic Top Chart despite being one of the most popular movie genres).

Top Romance Movies

In [75]:
build_chart('Romance').head(15)

title  year  vote_count  vote_average  \
10309  Dilwale Dulhania Le Jayenge  1995         661             9   
351                   Forrest Gump  1994        8147             8   
876                        Vertigo  1958        1162             8   
40248                   Your Name.  2016        1030             8   
883               Some Like It Hot  1959         835             8   
1132               Cinema Paradiso  1988         834             8   
19900                     Paperman  2012         734             8   
37860                  Sing Street  2016         669             8   
882                  The Apartment  1960         498             8   
38715               The Handmaiden  2016         453             8   
3189                   City Lights  1931         444             8   
24885             The Way He Looks  2014         262             8   
45434               In a Heartbeat  2017         146             8   
1639                       Titanic  1997        7770             7   
19730      Silver Linings Playbook  2012        4840             7   

       popularity        wr  
10309   34.457024  8.565285  
351     48.307194  7.971357  
876     18.208220  7.811667  
40248   34.461252  7.789489  
883     11.845107  7.745154  
1132    14.177005  7.744878  
19900    7.198633  7.713951  
37860   10.672862  7.689483  
882     11.994281  7.599317  
38715   16.727405  7.566166  
3189    10.891524  7.558867  
24885    5.711274  7.331363  
45434   20.821780  7.003959  
1639    26.889070  6.981546  
19730   14.488111  6.970581

The top romance movie according to our metrics is Dilwale Dulhania Le Jayenge.